File: c/triangle/triangle.c
#include "triangle.h"

static bool is_valid(triangle_t triangle) {
  return triangle.a + triangle.b >= triangle.c &&
         triangle.b + triangle.c >= triangle.a &&
         triangle.c + triangle.a >= triangle.b;
}

bool is_equilateral(triangle_t triangle) {
  return triangle.a > 0 && is_valid(triangle) && triangle.a == triangle.b &&
         triangle.b == triangle.c;
}

bool is_isosceles(triangle_t triangle) {
  return is_valid(triangle) &&
         (triangle.a == triangle.b || triangle.b == triangle.c ||
          triangle.c == triangle.a);
}

bool is_scalene(triangle_t triangle) {
  return is_valid(triangle) && triangle.a != triangle.b &&
         triangle.b != triangle.c && triangle.c != triangle.a;
}

File: javascript/triangle/triangle.js
export class Triangle {
  constructor(...sides) {
    [this.a, this.b, this.c] = sides;
    this.sides = new Set(sides);
  }

  get isValid() {
    return this.a + this.b >= this.c && this.b + this.c >= this.a && this.a + this.c >= this.b;
  }

  get isEquilateral() {
    return this.isValid && this.sides.size === 1 && !this.sides.has(0);
  }

  get isIsosceles() {
    return this.isValid && this.sides.size <= 2;
  }

  get isScalene() {
    return this.isValid && this.sides.size === 3;
  }
}
